### PR TITLE
HPCC-18934 Fix Windows build broken by libbase58 use of C99 extensions

### PR DESCRIPTION
**Pull request by afishbeck**
Also add cppunit test for libbase58 b58enc().

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by afishbeck** [comment]
@ghalliday please review

File: testing/unittests/unittests.cpp [hunk 1]
```diff
@@ -569,5 +570,58 @@ class StringBufferTest : public CppUnit::TestFixture
...
+
+        //hex 005A1FC5DD9E6F03819FCA94A2D89669469667F9A074655946
+        base64ToBase58(s.clear(), "AFofxd2ebwOBn8qUotiWaUaWZ/mgdGVZRg==");
+        ASSERT(streq(s, "19DXstMaV43WpYg4ceREiiTv2UntmoiA9j"));
+
+        //This input causes the loop iteration counter to go negative
+        //hex 00CEF022FA
+        base64ToBase58(s.clear(), "AM7wIvo=");
+        ASSERT(streq(s, "16Ho7Hs"));
+    }
```
Comment (ghalliday) [inline testing/unittests/unittests.cpp]: Good to see some tests!.

File: testing/unittests/unittests.cpp [hunk 2]
```diff
@@ -569,5 +570,58 @@ class StringBufferTest : public CppUnit::TestFixture
...
+    CPPUNIT_TEST_SUITE( Base58Test  );
+        CPPUNIT_TEST(testEncode);
+    CPPUNIT_TEST_SUITE_END();
+
+    void testEncode()
+    {
+        StringBuffer s;
+
+        //short string
+        ASSERT(streq(textToBase58(s.clear(), "1"), "r"));
```
Comment (ghalliday) [inline testing/unittests/unittests.cpp]: Good test.
Worth testing large character values as well.  e.g. "\xff", which unfortunately fails.

Comment (ghalliday) [reply testing/unittests/unittests.cpp]: Also good to test on a zero length string - which returns false

File: system/libbase58/base58.c [hunk 3]
```diff
@@ -38,7 +39,9 @@ bool b58tobin(void *bin, size_t *binszp, const char *b58, size_t b58sz)
     const unsigned char *b58u = (const unsigned char *)b58;
     unsigned char *binu = (unsigned char *)bin;
     size_t outisz = (binsz + 3) / 4;
-    uint32_t outi[outisz];
+    std::unique_ptr<uint32_t[]> outi(new uint32_t[outisz]);
```
Comment (ghalliday) [inline system/libbase58/base58.c]: This will be less efficient that using alloca.  I doubt it is significant, but how frequently are these functions due to be called?

**Review by ghalliday — COMMENTED** [review]
One general question - how long are the strings that this encoding is applied to?  From reading the code it will be O(N^2) in the number of bytes being encoded - which will be fine for short strings, but could get very expensive for long strings.

**Comment by ghalliday** [comment]
@afishbeck see comments

**Comment by afishbeck** [comment]
@ghalliday for my use these will always be small blocks generating the equivalent of a GUID type number or even smaller.  But I can change to alloca and add a couple more tests.

File: testing/unittests/unittests.cpp (hunk 2 above)
Comment (afishbeck) [reply testing/unittests/unittests.cpp]: I'll add a couple more tests, and see about the \xff failure and weather it matters.

File: system/libbase58/base58.c (hunk 3 above)
Comment (afishbeck) [reply system/libbase58/base58.c]: I thing the biggest use would be when LN implements there internal global transaction IDs, if one is passed to a roxie query that calls SOAPCALL or HTTPCALL out.  A rather short base58 id would be generated for each call out.  I'll change to alloca in any case.

**Comment by afishbeck** [comment]
@ghalliday updated, please re-review.

Note that we have an internal algorithm for global ids as well.  It will soon be open sourced and I'll look at the base58 part, and perhaps isolate it and use that if it's more efficient, if not perhaps optimize it first.

File: testing/unittests/unittests.cpp [hunk 4]
```diff
@@ -569,5 +570,71 @@ class StringBufferTest : public CppUnit::TestFixture
...
+
+        //This input causes the loop iteration counter to go negative
+        //hex 00CEF022FA
+        base64ToBase58(s.clear(), "AM7wIvo=");
+        ASSERT(streq(s, "16Ho7Hs"));
+
+        //empty input
+        MemoryBuffer mb;
+        size_t b58Length = 0;
+        ASSERT(!b58enc(s.reserve(b58Length), &b58Length, mb.toByteArray(), mb.length()));
```
Comment (ghalliday) [inline testing/unittests/unittests.cpp]: Is it correct that this fails?  When I commented that it returned false I wasn't convinced that was the correct result.

File: system/libbase58/base58.c [hunk 5]
```diff
@@ -38,7 +40,7 @@ bool b58tobin(void *bin, size_t *binszp, const char *b58, size_t b58sz)
     const unsigned char *b58u = (const unsigned char *)b58;
     unsigned char *binu = (unsigned char *)bin;
     size_t outisz = (binsz + 3) / 4;
-    uint32_t outi[outisz];
+    uint32_t *outi = (uint32_t*)alloca(outisz);
```
Comment (ghalliday) [inline system/libbase58/base58.c]: This is not equivalent.  It needs outisz * sizeof(uint32_t).  Why the tests didn't catch it?

File: testing/unittests/unittests.cpp [hunk 6]
```diff
@@ -569,5 +570,71 @@ class StringBufferTest : public CppUnit::TestFixture
 CPPUNIT_TEST_SUITE_REGISTRATION( StringBufferTest );
 CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( StringBufferTest, "StringBufferTest" );
 
+StringBuffer &mbToBase58(StringBuffer &s, const MemoryBuffer &data)
+{
+    size_t b58Length = data.length() * 2 + 1;
+
+    if (b58enc(s.reserve(b58Length), &b58Length, data.toByteArray(), data.length()) && b58Length > 1)
```
Comment (ghalliday) [inline testing/unittests/unittests.cpp]: Should this fail if b58enc returns false, or clear s(), otherwise you might have examples which accidentally work.

**Review by ghalliday — COMMENTED** [review]
A couple more comments

**Comment by ghalliday** [comment]
@afishbeck

File: system/libbase58/base58.c (hunk 5 above)
Comment (afishbeck) [reply system/libbase58/base58.c]: This code is unused by us, so my tests left it out, I will fix and add tests.

File: testing/unittests/unittests.cpp (hunk 4 above)
Comment (afishbeck) [reply testing/unittests/unittests.cpp]: I didn't think worth changing the 3rd party code as unlikely to be used this way, but I can take a look.

**Comment by afishbeck** [comment]
@ghalliday updated.  Some changes and also new tests for the decode routine.

File: testing/unittests/unittests.cpp [hunk 7]
```diff
@@ -124,7 +125,7 @@ int main(int argc, char* argv[])
     bool includeAll = false;
     bool verbose = false;
     bool list = false;
-    bool useDefaultLocations = true;
+    bool useDefaultLocations = false;
```
Comment (ghalliday) [inline testing/unittests/unittests.cpp]: @afishbeck was this change intentional?

**Comment by ghalliday** [comment]
one minor question, otherwise it looks ready to merge.

File: testing/unittests/unittests.cpp (hunk 7 above)
Comment (afishbeck) [reply testing/unittests/unittests.cpp]: @ghalliday oops that was to simplify my testing.  I'll remove.

**Comment by afishbeck** [comment]
@ghalliday removed spurious change.
